### PR TITLE
[Serializer] fixed fix encoding of cache keys with anonymous classes

### LINKED ISSUE #36589: [Serializer] Cache error normalizing anonymous class
**Symfony version(s) affected**: 4.4.7

**Description**  
During phpunit tests with `APP_DEBUG=1`, an anonymous class can be normalized by the `NormalizerInterface`, however, with `APP_DEBUG=0` the following error is produced:

```
Symfony\Component\Cache\Exception\InvalidArgumentException: Cache key "class@anonymous/app/tests/AnonymousClassTest.php:16$9c" contains reserved characters "{}()/\@:".
```

<details>
<summary>Full stack trace</summary>

```
Symfony\Component\Cache\Exception\InvalidArgumentException: Cache key "class@anonymous/app/tests/AnonymousClassTest.php:16$9c" contains reserved characters "{}()/\@:".

/app/vendor/symfony/cache/CacheItem.php:177
/app/vendor/symfony/cache/Traits/AbstractTrait.php:281
/app/vendor/symfony/cache/Traits/AbstractAdapterTrait.php:44
/app/vendor/symfony/cache/Adapter/PhpArrayAdapter.php:122
/app/vendor/symfony/serializer/Mapping/Factory/CacheClassMetadataFactory.php:50
/app/vendor/symfony/serializer/Mapping/ClassDiscriminatorFromClassMetadata.php:50
/app/vendor/symfony/serializer/Normalizer/AbstractObjectNormalizer.php:269
/app/vendor/symfony/serializer/Normalizer/AbstractObjectNormalizer.php:164
/app/vendor/symfony/serializer/Serializer.php:152
/app/tests/AnonymousClassTest.php:19
```

</details>

**How to reproduce**  

The error can be reproduced with the following test case:

```php
<?php

namespace App\Tests;

use Symfony\Bundle\FrameworkBundle\Test\KernelTestCase;
use Symfony\Component\Serializer\Normalizer\NormalizerInterface;

final class AnonymousClassTest extends KernelTestCase
{
    public function testSerializeAnonymousClass(): void
    {
        static::bootKernel();

        $normalizer = static::$container->get(NormalizerInterface::class);

        $class = new class() {
        };

        $normalizer->normalize($class);

        $this->addToAssertionCount(1);
    }
}
```

<details>
<summary>Successful test run with APP_DEBUG=1</summary>

```
$ APP_DEBUG=1 vendor/bin/phpunit tests/AnonymousClassTest.php 
PHPUnit 8.5.4 by Sebastian Bergmann and contributors.

.                                                                   1 / 1 (100%)

Time: 466 ms, Memory: 20.00 MB

OK (1 test, 1 assertion)
```

</details>

<details>
<summary>Failed test run with APP_DEBUG=0</summary>

```
$ APP_DEBUG=0 vendor/bin/phpunit tests/AnonymousClassTest.php 
PHPUnit 8.5.4 by Sebastian Bergmann and contributors.

E                                                                   1 / 1 (100%)

Time: 216 ms, Memory: 18.00 MB

There was 1 error:

1) App\Tests\AnonymousClassTest::testSerializeAnonymousClass
Symfony\Component\Cache\Exception\InvalidArgumentException: Cache key "class@anonymous/app/tests/AnonymousClassTest.php:16$9c" contains reserved characters "{}()/\@:".

/app/vendor/symfony/cache/CacheItem.php:177
/app/vendor/symfony/cache/Traits/AbstractTrait.php:281
/app/vendor/symfony/cache/Traits/AbstractAdapterTrait.php:44
/app/vendor/symfony/cache/Adapter/PhpArrayAdapter.php:122
/app/vendor/symfony/serializer/Mapping/Factory/CacheClassMetadataFactory.php:50
/app/vendor/symfony/serializer/Mapping/ClassDiscriminatorFromClassMetadata.php:50
/app/vendor/symfony/serializer/Normalizer/AbstractObjectNormalizer.php:269
/app/vendor/symfony/serializer/Normalizer/AbstractObjectNormalizer.php:164
/app/vendor/symfony/serializer/Serializer.php:152
/app/tests/AnonymousClassTest.php:19

ERRORS!
Tests: 1, Assertions: 0, Errors: 1.
```

</details>

**Possible Solution**  

It appears that an issue similar to this has appeared a few times in the past (linked issued below) and it was fixed with something like this in `PropertyAccessor::getReadAccessInfo` and `PropertyAccessor::getWriteAccessInfo`:

```
$key = false !== strpbrk($key = $class.'..'.$property, '{}()/@:') ? rawurlencode($key) : $key;
```

But this issue seems to be unrelated to the PropertyAccessor, so I adding the change to `CacheClassMetadataFactory::getMetadataFor` and that makes the test pass successfully with `APP_DEBUG=0`:

```diff
public function getMetadataFor($value)
{
    $class = $this->getClass($value);
    // Key cannot contain backslashes according to PSR-6
    $key = strtr($class, '\\', '_');
+    $key = false !== strpbrk($key, '{}()/@:') ? rawurlencode($key) : $key;

    $item = $this->cacheItemPool->getItem($key);

```

Not sure if this is the correct place to fix the issue, or if it has knock on affects elsewhere.

**Additional context**  

I found a few older issues which may be related:
* #23136
* #36565 
* #29329 
* #29293 
* #26313

### PR DESCRIPTION
**Pull request by michaelzangerle**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #36589 
| License       | MIT
| Doc PR        | -

### CODE REVIEW

File: src/Symfony/Component/Serializer/Mapping/Factory/CacheClassMetadataFactory.php [hunk 1]
```diff
@@ -46,6 +46,7 @@ public function getMetadataFor($value)
         $class = $this->getClass($value);
         // Key cannot contain backslashes according to PSR-6
         $key = strtr($class, '\\', '_');
+        $key = false !== strpbrk($key, '{}()/@:') ? rawurlencode($key) : $key;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Mapping/Factory/CacheClassMetadataFactory.php:49]: What about replacing the previous line by this instead?
`$key = strtr($class, '\\:', '__');`

Comment (michaelzangerle) [reply src/Symfony/Component/Serializer/Mapping/Factory/CacheClassMetadataFactory.php:49]: @nicolas-grekas  Applying your suggestion will result in the following and still fail
```
class.anonymous/symfony/src/Symfony/Component/Serializer/Tests/Mapping/Factory/CacheMetadataFactoryTest.php-69$b7
```

The way it is now it results in 
```
class%40anonymous%00%2Fsymfony%2Fsrc%2FSymfony%2FComponent%2FSerializer%2FTests%2FMapping%2FFactory%2FCacheMetadataFactoryTest.php%3A69%24b7
```

How about we do it similar to the change [here](https://github.com/symfony/symfony/pull/29313/files#diff-3e45cf556c0330623c0167841d303dffR511) and replace both lines with

```
$key = false !== strpbrk($class, '{}()/@:') ? rawurlencode($class) : $class;
```

Comment (nicolas-grekas) [reply src/Symfony/Component/Serializer/Mapping/Factory/CacheClassMetadataFactory.php:49]: > How about we do it similar to the change here and replace both lines with

works for me

Comment (michaelzangerle) [reply src/Symfony/Component/Serializer/Mapping/Factory/CacheClassMetadataFactory.php]: The last change works for the serializer component but affects a test case in the FrameworkBundle. The keys used [here](https://github.com/symfony/symfony/blob/v4.4.11/src/Symfony/Bundle/FrameworkBundle/Tests/CacheWarmer/SerializerCacheWarmerTest.php#L40) do not match anymore. I am not really sure about the impact this change has. Should I adjust the test case or revert it back to the first version which did not cause this? :thinking:

**Comment by michaelzangerle** [comment]
Adjusted test case in the framework bundle to match new cache key escaping. Tests for serializer and framework bundle run through locally and in AppVeyor but fail on Travis for some reason. One error on travis is probably related to the change but I can't reproduce it locally as it should be fixed with the change. The other is a redis error according to the error message. Any advice?

File: src/Symfony/Component/Serializer/Mapping/Factory/CacheClassMetadataFactory.php [hunk 2]
```diff
@@ -44,8 +44,8 @@ public function __construct(ClassMetadataFactoryInterface $decorated, CacheItemP
     public function getMetadataFor($value)
     {
         $class = $this->getClass($value);
-        // Key cannot contain backslashes according to PSR-6
-        $key = strtr($class, '\\', '_');
+        // Key cannot contain backslashes, colons, brackets, slashes and @ according to PSR-6
+        $key = false !== strpbrk($class, '\{}()/@:') ? rawurlencode($class) : $class;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Mapping/Factory/CacheClassMetadataFactory.php:48]: one last note: let's optimise for the most likely case: the backslash is going to be here 99% of the time.
`$key = rawurlencode($class);` should be the best here.

Comment (michaelzangerle) [reply src/Symfony/Component/Serializer/Mapping/Factory/CacheClassMetadataFactory.php]: @nicolas-grekas Thanks. I applied suggested change.

**Comment by nicolas-grekas** [comment]
Thank you @michaelzangerle.

**Pull request merged**